File: backend/internal/models/workflow.go
package models

import (
	"time"
)

// NodeType represents the type of a workflow node
type NodeType string

const (
	NodeTypeHTTPRequest NodeType = "httpRequest"
	NodeTypeTransform   NodeType = "transform"
	NodeTypeCondition   NodeType = "condition"
	NodeTypeTrigger     NodeType = "trigger"
	NodeTypeWebhook     NodeType = "webhook"
)

// HTTPMethod represents HTTP request methods
type HTTPMethod string

const (
	MethodGET    HTTPMethod = "GET"
	MethodPOST   HTTPMethod = "POST"
	MethodPUT    HTTPMethod = "PUT"
	MethodDELETE HTTPMethod = "DELETE"
	MethodPATCH  HTTPMethod = "PATCH"
)

// Node represents a workflow node
type Node struct {
	ID       string                 `json:"id"`
	Type     NodeType               `json:"type"`
	Position Position               `json:"position"`
	Data     NodeData               `json:"data"`
}

// Position represents the position of a node on the canvas
type Position struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// NodeData contains the node's data
type NodeData struct {
	Label   string                 `json:"label"`
	Config  map[string]interface{} `json:"config"`
	Outputs map[string]interface{} `json:"outputs,omitempty"`
}

// Edge represents a connection between two nodes
type Edge struct {
	ID     string `json:"id"`
	Source string `json:"source"`
	Target string `json:"target"`
	Type   string `json:"type"`
}

// Workflow represents a complete workflow
type Workflow struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	Description string    `json:"description,omitempty"`
	Nodes       []Node    `json:"nodes"`
	Edges       []Edge    `json:"edges"`
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
	Version     int       `json:"version"`
}

// ExecutionStatus represents the status of an execution
type ExecutionStatus string

const (
	StatusRunning   ExecutionStatus = "running"
	StatusCompleted ExecutionStatus = "completed"
	StatusFailed    ExecutionStatus = "failed"
)

// ExecutionResult represents the result of a node execution
type ExecutionResult struct {
	NodeID   string                 `json:"nodeId"`
	Status   string                 `json:"status"`
	Output   map[string]interface{} `json:"output,omitempty"`
	Error    string                 `json:"error,omitempty"`
	Duration int64                  `json:"duration,omitempty"` // milliseconds
}

// WorkflowExecution represents a workflow execution instance
type WorkflowExecution struct {
	ID          string            `json:"id"`
	WorkflowID  string            `json:"workflowId"`
	Status      ExecutionStatus   `json:"status"`
	StartedAt   time.Time         `json:"startedAt"`
	CompletedAt *time.Time        `json:"completedAt,omitempty"`
	Results     []ExecutionResult `json:"results"`
}

// HTTPRequestConfig represents configuration for HTTP request node
type HTTPRequestConfig struct {
	URL     string            `json:"url"`
	Method  HTTPMethod        `json:"method"`
	Headers map[string]string `json:"headers,omitempty"`
	Body    interface{}       `json:"body,omitempty"`
}

// TransformConfig represents configuration for transform node
type TransformConfig struct {
	Code     string `json:"code"`
	Language string `json:"language"` // "javascript" or "jq"
}

// ConditionConfig represents configuration for condition node
type ConditionConfig struct {
	Conditions []Condition `json:"conditions"`
	Combinator string      `json:"combinator"` // "AND" or "OR"
}

// Condition represents a single condition
type Condition struct {
	Field    string      `json:"field"`
	Operator string      `json:"operator"` // "equals", "notEquals", "contains", "greaterThan", "lessThan"
	Value    interface{} `json:"value"`
}

File: frontend/src/types/workflow.ts
import { Node, Edge } from 'reactflow';
import { z } from 'zod';

// Node Type Enums
export enum NodeType {
  HTTP_REQUEST = 'httpRequest',
  TRANSFORM = 'transform',
  CONDITION = 'condition',
  TRIGGER = 'trigger',
  WEBHOOK = 'webhook',
}

// Zod Schemas for validation
export const httpRequestSchema = z.object({
  url: z.string().url(),
  method: z.enum(['GET', 'POST', 'PUT', 'DELETE', 'PATCH']),
  headers: z.record(z.string()).optional(),
  body: z.any().optional(),
  authentication: z
    .object({
      type: z.enum(['none', 'basic', 'bearer', 'apiKey']),
      credentials: z.record(z.string()),
    })
    .optional(),
});

export const transformSchema = z.object({
  code: z.string(),
  language: z.enum(['javascript', 'jq']),
});

export const conditionSchema = z.object({
  conditions: z.array(
    z.object({
      field: z.string(),
      operator: z.enum(['equals', 'notEquals', 'contains', 'greaterThan', 'lessThan']),
      value: z.any(),
    })
  ),
  combinator: z.enum(['AND', 'OR']),
});

export const webhookSchema = z.object({
  path: z.string(),
  method: z.enum(['GET', 'POST', 'PUT', 'DELETE']),
});

// TypeScript Types
export type HttpRequestConfig = z.infer<typeof httpRequestSchema>;
export type TransformConfig = z.infer<typeof transformSchema>;
export type ConditionConfig = z.infer<typeof conditionSchema>;
export type WebhookConfig = z.infer<typeof webhookSchema>;

export type NodeConfig =
  | HttpRequestConfig
  | TransformConfig
  | ConditionConfig
  | WebhookConfig;

export interface WorkflowNode extends Node {
  type: NodeType;
  data: {
    label: string;
    config: NodeConfig;
    outputs?: any;
  };
}

export interface WorkflowEdge extends Edge {}

export interface Workflow {
  id: string;
  name: string;
  description?: string;
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  createdAt: string;
  updatedAt: string;
  version: number;
}

export interface ExecutionResult {
  nodeId: string;
  status: 'success' | 'error' | 'pending';
  output?: any;
  error?: string;
  duration?: number;
}

export interface WorkflowExecution {
  id: string;
  workflowId: string;
  status: 'running' | 'completed' | 'failed';
  startedAt: string;
  completedAt?: string;
  results: ExecutionResult[];
}

export interface ApiError {
  message: string;
  code: string;
  details?: any;
}
